src/gc/proxy.d

```diff
@@ -28,32 +28,67 @@ private
     extern (C) void thread_init();
     extern (C) void thread_term();
 
+    import core.internal.spinlock;
+    static SpinLock instanceLock;
+
     __gshared GC instance;
     __gshared GC proxiedGC; // used to iterate roots of Windows DLLs
-
 }
 
+alias GCInitNoThrow = void function() nothrow @nogc;
 
 extern (C)
 {
 
     void gc_init()
     {
-        config.initialize();
-        ManualGC.initialize(instance);
-        ConservativeGC.initialize(instance);
-        if (instance is null)
-        {
-            import core.stdc.stdio : fprintf, stderr;
-            import core.stdc.stdlib : exit;
+        import core.atomic : atomicLoad, atomicStore;
+
+        auto pinstance = cast(shared(GC)*) &instance;
 
-            fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
-            exit(1);
+        if (atomicLoad(*pinstance) is null)
+        {
+            instanceLock.lock();
+            scope(exit) instanceLock.unlock();
+            {
+                if (atomicLoad(*pinstance) is null)
+                {
+                    GC auxInstance;
+                    config.initialize();
+                    ManualGC.initialize(auxInstance);
+                    ConservativeGC.initialize(auxInstance);
+
+                    if (auxInstance is null)
+                    {
+                        import core.stdc.stdio : fprintf, stderr;
+                        import core.stdc.stdlib : exit;
+
+                        fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
+                        instanceLock.unlock();
+                        exit(1);
+                    }
+
+                    atomicStore(*pinstance, cast(shared GC)auxInstance);
```

Comment:
Race condition?  setting `auxInstance` to "not null" should be the last thing done correct?  Otherwise another thread could come in, see that `auxInstance` is not null and then continue before `thread_init` has finished.   However, my comment is assuming that `thread_init` should be finished before another thread can start using the GC, is my assumption incorrect?

Reply:
Depends on the startup, there could be multiple external "C" threads calling `thread_attachThis` concurrently.

Reply:
@MartinNowak so what's the right course of action here?

Reply:
The mutual dependency between gc_init and thread_init isn't properly resolved.
```d
if (atomicLoad(*pinstance) is null))
{
    instanceLock.lock();
    scope (exit) instanceLock.unlock();
    if (atomicLoad(*pinstance) is null))
    {
        // init
        // ...
        atomicStore(*pinstance, cast(shared GC)auxInstance);

        // thread_init <- should be moved before storing pInstance or some other thread could start to use the GC before thread_init finishes
    }
}
```
The problem is that thread_init depends on the GC, but any GC usage depends on thread_init (for suspending).
Let's figure out whether we can avoid the GC for Thread.getThis, if not then pass `auxInstance` directly to thread_init.

Reply:
And BTW, use weaker (and much cheaper) orderings.
```d
if (atomicLoad!(MemoryOrder.acq)(*pInstance) is null)
{
    // lock...
    if (atomicLoad!(MemoryOrder.acq)(*pInstance) is null)
    {
        // init...
        atomicStore!(MemoryOrder.rel)(*pInstance, auxInstance);
    }
}

Reply:
The second `atomidLoad` can be `raw` correct? (see https://github.com/dlang/phobos/pull/6063)
```D
if (atomicLoad!(MemoryOrder.acq)(*pInstance) is null)
{
    // lock...
    if (atomicLoad!(MemoryOrder.raw)(*pInstance) is null)
    {
        // init...
        atomicStore!(MemoryOrder.rel)(*pInstance, auxInstance);
    }
}
```

Reply:
> pass `auxInstance` directly to thread_init

This seems like the correct answer.

Reply:
@MartinNowak let's make it most conservative and optimize in a subsequent step thx

Reply:
This is nastier than it looks. 
```thread_init``` calls ```thread_attachThis``` which calls ```GC.disable``` and ```GC.enable```.
https://github.com/dlang/druntime/blob/master/src/core/thread.d#L2138

While ```gc_init``` has not yet finished, the spinlock is still locked. Calling ```disable``` will redirect to ```gc_init``` (with the ```instance``` still ```null``` because ```store``` hasn't been done yet) -> deadlock.

Reply:
That works for ```gc_disable```/```gc_enable```.
However, the code then creates an object ```new Thread``` which will become a call to ```_d_newclass``` in ```lifetime.d```. 
https://github.com/dlang/druntime/blob/master/src/rt/lifetime.d#L93

This function calls ```GC.malloc```. I can't pass the instance there and the code hangs on the spinlock (as ```GC.malloc``` tries to call ```gc_init```).

A slightly ugly workaround is to create a static ```tlsInstance``` for the thread doing the initialization and use this one, if available, in ```gc_enable```, ```gc_disable``` and ```gc_malloc```.

Reply:
`_d_newclass` should forward to a function that accepts a gc instance, and then instead of calling `new Thread`, call the function directly (with the `auxInstance` of course).